### PR TITLE
add init version of Explore page

### PR DESCRIPTION
**Pull request by Simek**
# Why

Currently the Directory do not offer any package discoverability focus features, which we think could be a nice addition (to the browsering and searching through the packages). 

**The implementation is simple, so please treat this like Draft/Beta.**

This PR includes following changes:
* new Explore page which highlights several libraries mainly based on the NPM download count
  * monthly download count is compared to the last week stats and basic filtering of inactive and deprecated libs is performed - this produce a `popularity` metrics which represents the grow or shrink of download numbers in last period in comparison to the monthly value divided by `4` (the few rest days in month will serve as a small buffer to stabilize the results)
  * Explore page is divided to the sections to give core and each out-of-tree platform equal space and an equal opportunity to for platform specific libraries to appear on the list
  * Explore sections `Library` component tiles are stripped of the most metadata, just to shift focus on the library and fit two of them in row on non-mobile devices (I have opted-out from using CSS grids because they were some minor display issues across the browsers) 
  * probably after adding the last week stats the `npm` object structure in data can be refactored, but I have opted-out from this for now, just to focus on the draft, this can be improved at any time

### Preview

For most recent screenshots, scroll down the thread.

### Note

* ~I had to use inline styles for the `<sup>` tag, because passing styles object was causing an error (source of error is probably `@expo/html-elements` or `react-native-web`, I will investigate this further)~
* This will be run with each update of data file, but I'm not sure how often the libraries will shift position or change completely, (currently the biggest problem will be desktop platforms - macOS and Windows, because there is not many libraries supporting those platforms)
* in the future we should add also `tvOS` and `Development Tools` sections, when there will be more libraries fitting those categories

# Checklist

- [X] Documented in this PR how to use the feature or replicate the bug.
- [X] Documented in this PR how you fixed or created the feature.

### CODE REVIEW

**Comment by Simek** [comment]
In the latest commit I have added:
* basic popularity mark, mainly for debug purposes at this time (popularity score listed in `aria-label`)
* note leading to the Directory main page at the end of Explore page
* minor CSS fixes and tweaks related to the Explore page and Library component

**Comment by brentvatne** [comment]
hi @simek! thanks for putting this together and continuing to push the directory forward 👏 

- the colored indicators on the top right of cards are kind of unclear to me without reading the code, maybe we should explore other options for communicating relative popularity
- i think i would prefer to see the list of sections mirror the platforms listed in the footer. this means we should remove "core platforms" and "expo" from the list.
- a very different idea entirely but: what if we had an option to show smaller cards on our homepage (like those in the mockup) so we could fit more libraries on one screen, and then added an option to sort by trending? we could make the `/explore` path apply the small card and sort by trending options automatically, so the result for the user would be similar to the mockup and users could take advantage of the other features we already have like filtering by platform and so on.
- users could also bookmark the page with filters they care about, like trending libraries for windows, rather than having to scroll own to the bottom of the page each time with the current design

**Comment by Simek** [comment]
Thank you for the feedback @brentvatne! 

> the colored indicators on the top right of cards are kind of unclear to me without reading the code, maybe we should explore other options for communicating relative popularity

I agree, as stated in the first comments, the indicators serves currently the debug/testing purpose, so I can see the calculated popularity (without accessing `data.json`). It would be more useful after the deployment, but for me we can disable the indicators for now.

> i think i would prefer to see the list of sections mirror the platforms listed in the footer. this means we should remove "core platforms" and "expo" from the list.

I was thinking about that too, removing "Core platforms" will cause us missing the "trending" libraries which supports both platforms due to how currently iOS and Android filters works (`lib.android && !lib.ios`). Changing the filters and removing negation leads to very similar results in both Android and iOS sections (at least from my testing, this was the solution which which yields the best results for me so far). More test and experiments might reveal better approach. The section design was mainly data driven 😉 

> a very different idea entirely but: what if we had an option to show smaller cards on our homepage (like those in the mockup) so we could fit more libraries on one screen, and then added an option to sort by trending? we could make the /explore path apply the small card and sort by trending options automatically, so the result for the user would be similar to the mockup and users could take advantage of the other features we already have like filtering by platform and so on.

It's quite interesting idea, I think that after tweaking the algorithm, the `popularity` score cloud just become another sort order option. Also we can try to add 1 or 2 columns display option togglers for the main Directory browser, that could be nice also, but I would like to maintain the separation between the browser and Explore/Discovery page. I think that it give us a playground, a place where we can experiment, without affecting the whole directory, and I have a feeling that in a long term those features will gradually move in the opposite directions. 

> users could also bookmark the page with filters they care about, like trending libraries for windows, rather than having to scroll own to the bottom of the page each time with the current design

I was thinking about this aspect too. As a basic improvement we can add the hash links, but generally, in a long term, it would be nice to allow per-user customization of the Explore page. Ideally I would like to see login with GitHub option, which should enable:
* sending/updating libs from the popup form in Directory
* loading/saving the filters and sort settings
* customizing the Explore page (show/hide platform sections, select count etc.)

I can also agree that generally this page might seems a bit rough and predefined, but at the beginning of work I would like to focus more on the popularity algorithm itself, score calculations and data related improvements, rather than the UI/UX of the Explore page. 

One of purposes for this page, which I have in mind was, that this page can serve as a source of daily set of libraries which user can check at the beginning or end of the day and do not sink much time into that and this thread, which I'm subscribed for a long time, was one of my inspirations - https://github.com/vitalets/github-trending-repos/issues/5#issuecomment-733326801. I was using it mainly before the GitHub "Explore" tabs emerges,  same with - https://taoalpha.github.io/npm-trending/ - which provides the nice data, but no ability to reduce the scope or filter out some libraries.

**Comment by Simek** [comment]
### Changes since last feedback:
* `PopularityMark` has been refactored to output more intuitive/meaningful data for the users, the score in bracket can be disabled after the algorithm testing period, and the level/grade captions definitely can be improved, those were just my quick ideas
* popularity mark visibility has been splitted out `skipMeta` flag, so now `Library` component can accept also `showPopularity` flag
* on the Explore page generic "browse the directory" note has been replaced with parametrized browser link for each of sections, so users should be able to continue the exploration of libraries for the desired platform even easier

### Updated Preview
React Native Directory (11)

**Comment by Simek** [comment]
List of the most important changes since last comment:
* popularity calculations has been moved to the `build&score` data script
* the Explore page libraries filters has been refactored to the popularity score penalties and moved to the initial calculation
* `Popularity Gain` sorting has been added to the API and utilized in directory and on the Explore page
* various TS types has been fixed and `QueryOrder` type has been added

### Preview after latest data fetch and updates
React Native Directory (14)

**Comment by Simek** [comment]
### Preview with Trending Topics experiment
React Native Directory (15)

**Comment by Simek** [comment]
I have made a quite big refactor of Explore experience, divining the whole page to the three sections (inspired by GitHub Explore sections):
* `/explore` - containing short lists of top, popular libraries per platform - https://github.com/explore
* `/topics` - containing the most trending topics across all the libraries in the Directory - https://github.com/topics
* `/trending` - containing the packages list ordered by popularity and filters box, so users can play and shape the list for themselves - https://github.com/trending

To make testing process easier, I have uploaded regenerated and Explore page ready dataset:
* https://gist.github.com/Simek/f48aa903adc99c62507ebd3527987b7b

I have also merged the latest `master` changes into this PR, so the changes have be radically reduced, which should help others to hop into the mix or during the review.

File: components/Explore/ExploreNav.tsx
Comment (jonsamp) [inline components/Explore/ExploreNav.tsx:41]:
```suggestion
    paddingVertical: 40,
```

We could use more padding here. 
Screen Shot 2021-02-28 at 2 01 36 PM

File: scripts/fetch-npm-data.js
Comment (jonsamp) [inline scripts/fetch-npm-data.js:29]:
```suggestion
```

**Review by jonsamp — APPROVED** [review]
Overall, I really like this addition to the directory. Thank you so much for putting forth so much effort and refining this. If you do the things I suggested above, you have my approval! If you'd like more feedback/need more clarity, I'm happy to review more.

File: components/Header.tsx
Comment (Simek) [reply components/Header.tsx:50]: I will remove it, no problem. 🙂

File: pages/explore.tsx
Comment (Simek) [reply pages/explore.tsx:194]: Sure, don't have a strong preference in there, can agree that usual, left align fits better the general Directory style. 😉

File: pages/topics.tsx
Comment (Simek) [reply pages/topics.tsx:72]: TLDR: I agree, and I will remove the Topics subpage for now. 🙂 

The "Tag cloud" started as an component at the bottom of "Explore" subpage, as a one of many sections in there there was no need to output quite large amount of tags so I have tweaked the acceptance/filter param, so only few tags (6-8) were present (https://github.com/react-native-community/directory/pull/509/files#diff-ffa524df88e52d014c665a25578de7e7bb7ed33bbb67d0581a4618478c85eb23R38). 

After moving "Topics" to the separate subpage I have reduced the popularity threshold, so more tags can appear on the cloud. As you spotted this reduced the overall quality and reduced the usefulness of tags, also the change do NOT fix the page emptiness issue.

So I think it's fair to remove this subpage for now, maybe in the future we will able to work on more robust and meaningful for the users implementation of topics. 😉

File: components/Library/PopularityMark.tsx
Comment (Simek) [reply components/Library/PopularityMark.tsx:39]: Awesome, thank you for the colors tweaks @jonsamp! I have applied all of them, but slightly tweaked the shade of orange. 🧡

**Comment by Simek** [comment]
Also it's quite important to note that at this moment the "Trending" page did not include many tweaks and advantages (besides the  `PopularityMark`) in comparison to the base libraries list with "Popularity gain" sort option, but it probably will evolve a bit with the time and feedback from users.

The difference can grow even more after introducing GitHub login to the Directory, as Bren suggested in one of the comments.

ATM the "Trending" explore page do not offer pagination and it's hardcoded to return only first 20 matches, I would like to switch it to "unlimited" list, but  with additional hard coded popularity filter so libraries with popularity score below `0` won't show on that list, to further differentiate the both list, what do you think about this idea @jonsamp ?

**Comment by jonsamp** [comment]
> With value set to 0.025 (2.5) and with Explore test dataset the list contains 64 libraries without the filters.

I like this a lot! Let's do it.

Once you do that, let's get this PR merged :D

**Comment by Simek** [comment]
> Once you do that, let's get this PR merged :D

Commit is already pushed - https://github.com/react-native-community/directory/pull/509/commits/7f094e75ce3f4da17c29ec93c978e61105a71373 🙂 

Let me look at and implement API popularity number filter, so we don't choke API after merge.

**Comment by Simek** [comment]
New `minPopularity` filter has been added to the API, I have stick to the values format in the UI (range up to `100`) instead of passing floats (up to `1` - as represented in API data) just for the ease of use, but this might be confusing in a long run:
image

I have also updated the Query type and Explore pages to utilize this new filter.

**Comment by Simek** [comment]
@brentvatne I have refactored the weekly data fetch to recently added bulk queries, so now the impact of additional fetches should not be that visible. On localhost this have added only about a minute to the `data:update` script runtime.

If you don't have an additional request or suggestions this PR should be ready to review and merge! 🙂

**Review by brentvatne — APPROVED** [review]

**Pull request merged**
